src/activity_actor.cpp

```diff
@@ -2348,6 +2348,124 @@ std::unique_ptr<activity_actor> move_furniture_activity_actor::deserialize( Json
+{
+    // Cost to take an item from a container or map
+    const int obtain_cost = item.obtain_cost( who );
+    // Cost to move an item to a container, vehicle or the ground
+    const int move_cost = Pickup::cost_to_move_item( who, *item );
+    return obtain_cost + move_cost;
+}
+
+void insert_item_activity_actor::start( player_activity &act, Character &who )
+{
+    if( items.empty() ) {
+        debugmsg( "ACT_INSERT_ITEM was passed an empty list" );
+        act.set_to_null();
+    }
+
+    all_pockets_rigid = holster->contents.all_pockets_rigid();
+
+    const int total_moves = item_move_cost( who, items.front().first );
+    act.moves_left = total_moves;
+    act.moves_total = total_moves;
+}
+
+void insert_item_activity_actor::finish( player_activity &act, Character &who )
+{
+    bool success = false;
+    drop_location &holstered_item = items.front();
+    if( holstered_item.first ) {
+        item &it = *holstered_item.first;
+        if( !it.count_by_charges() ) {
+            if( holster->can_contain( it ) && ( all_pockets_rigid ||
+                                                holster.parents_can_contain_recursive( &it ) ) ) {
+
+                success = holster->put_in( it, item_pocket::pocket_type::CONTAINER,
+                                           /*unseal_pockets=*/true ).success();
+                if( success ) {
+                    handler.add_unsealed( holster );
+                    handler.unseal_pocket_containing( holstered_item.first );
+                    holstered_item.first.remove_item();
+                }
+
+            }
+        } else {
+            int charges = all_pockets_rigid ? holstered_item.second : std::min( holstered_item.second,
+                          holster.max_charges_by_parent_recursive( it ) );
+
+            if( charges > 0 && holster->can_contain_partial( it ) ) {
+                int result = holster->fill_with( it, charges,
+                                                 /*unseal_pockets=*/true,
+                                                 /*allow_sealed=*/true );
+                success = result > 0;
+
+                if( success ) {
+                    handler.add_unsealed( holster );
+                    handler.unseal_pocket_containing( holstered_item.first );
+                    it.charges -= result;
+                    if( it.charges == 0 ) {
+                        holstered_item.first.remove_item();
+                    }
+                }
+            }
+        }
+
+        if( !success ) {
+            who.add_msg_if_player(
+                string_format(
+                    _( "Could not put %1$s into %2$s, aborting." ),
+                    it.tname(), holster->tname() ) );
+        }
+    } else {
+        // item was lost, so just go to next item
+        success = true;
+    }
+
+    items.pop_front();
+    if( items.empty() || !success ) {
+        handler.handle_by( who );
+        act.set_to_null();
+        return;
+    }
```

Comment:
`handle_by` may ask the player to handle unsealed containers, so it should only be called when all items are moved. Otherwise if the containers are moved or contents are spilled, later invocation of this function will use invalid item locations.